docs/designs/2717-produce-tce-specific-releases.md

```diff
@@ -0,0 +1,510 @@
+* Push (upstream) OS images to relevant cloud providers
+* Generate Tanzu Kubernetes Release (TKr)
+* Push the TKr to an OCI repository
+
+The following diagram visually represents the above.
+
+![Diagram of build manifest assets being processed by
+CI](imgs-2717/build-manifest-ci.png)
+
+Packages (bundles pushed using `imgpkg`) are **not** pushed in this process.
+Instead, package maintainers are expected to have already pushed a given
+package. The build metadata above only contains a reference to the pushed
+package. With this, the CI can construct a package repository, which is a image
+holding references and metadata to the packages.
+
+The asset produced above will power cluster creation at runtime in the TKr.
+The TKr should **not** include build information such as specific packages used.
+Instead, it should reference what is relevant at runtime. This includes assets
+such as package repositories and host OS images.
+
+##### Schema
+
+The proposed schema of the build manifest is:
+
+```go
+type TanzuBuild struct {
+  // the version of this release and build
+  Version string
+  // packages to create a core package repo of
+  // created repo will be uploaded to an OCI repo
+  // uri of uploaded repo will be added to the generated TKR
+  CorePackages []MetaPackage `yaml:"corePackages"`
+  // packages to create a user-managed package repo of
+  // created repo will be uploaded to an OCI repo
+  // uri of uploaded repo will be added to the generated TKR
+  UserPackages []MetaPackage `yaml:"userPackages"`
+  // host images to add to the generated TKR
+  HostImages []HostImage `yaml:"hostImages"`
+  // k8s metadata added to the generated TKR
+  KubernetesMeta `yaml:"kubernetesMeta"`
+}
+
+type MetaPackage struct {
+  // name of the metapackage, e.g. contour
+  Name string
+  // each versioned package to make available
+  Packages []Package
+  // Embeds contents of package metadata:
+  // https://carvel.dev/kapp-controller/docs/latest/packaging/#package-metadata
+  PackageMetadata string  `yaml:"packageMetadata"`
+}
+
+type Package struct {
+  // fully qualified name of package, e.g. contour.dev.1.5.3
+  Name string
+  // location of the OCI bundle
+  ImageBundleURI string `yaml:"imageBundleUri"`
+  // Arbitrary set of options to add in the package CR
+  Options map[string]interface{}
+}
+
+type HostImage struct {
+  // name of the host image (ami-dfdal)
+  Name string
+  // the infrastructure provider it relates to (aws, azure, etc)
+  Provider string
+  // metadata relevant to the specific provider. (aws-region: us-west-2)
+  Metadata map[string]string
+}
+
+type KubernetesMeta struct {
+  // the version of Kubernetes this TKR represents
+  Version string
+  // the kubernetes components used in bootstrap
+  // these should match what is inside the host images
+  // otherwise kubeadm will pull down new images on each
+  // host during bootstrap
+  Components []ContainerImage
+}
+
+type ContainerImage struct {
```

Comment:
Are these listed as special case container images because the `kapp` controller isn't actually running at this point of cluster creation?